### PR TITLE
add mentor walkthrough steps

### PR DESCRIPTION
**Pull request by ZooeyMiller**
This is being merged to `mentor-walkthrough` not `master` as it only here for this code-along to be consistent in all countries and all future times it is taught. It is for mentors, *not* students os I do not think it should be in master.

This PR adds a series of `step-...` files which contain working examples of how the code should look at all of the different points of the code along. The end of each of these steps is also intended to be a logical point to bring everyone up to the same level in case anyone is lagging behind.

It also adds `node-code-along-walkthrough.md` which is a guide for mentors running the code along that they can study in advanced, and refer to when needed during the code along. 
#7

### CODE REVIEW

**Comment by jsms90** [comment]
@ZooeyMiller pulling in some help from @marisid here :+1: 

Try to be super descriptive with the PR title :smile:

File: src/handlers/handlers.js
Comment (eliasmalik) [inline src/handlers/handlers.js]: Would prefer not to plant the idea of throwing in a handler. I've been commenting during code reviews about this during every cohort. I can see you've used the proper approach in other steps, so perhaps this was an oversight?

File: src/router/router.js
Comment (eliasmalik) [inline src/router/router.js]: As above, RE: throwing

**Review by eliasmalik — CHANGES_REQUESTED** [review]
1. As I mentioned foundersandcoders/master-reference#456, I dislike the whole "keep notes/solutions on a different branch" approach. Much prefer having dotfiles or dotfolders for mentor notes. @des-des @bradreeder @jsms90 thoughts?

2. Do we really need the lock-files all the other stuff? Can't we just have the JS/HTML/CSS?

File: src/handlers/handlers.js
Comment (ZooeyMiller) [reply src/handlers/handlers.js]: Happy to change this

File: src/server.js
Comment (marisid) [inline src/server.js]: Worth highlighting the difference between `start` and `devStart` script tags and the fact that in a development environment you would like to see the changes you make to your app readily, hence `nodemon` is very useful.

File: src/server.js
Comment (marisid) [inline src/server.js]: Maybe add here that `install`, `start` and `test` are three common default tags that can be run directly by typing `npm <name-of-script-tag>`

File: src/handlers/handlers.js
Comment (ZooeyMiller) [reply src/handlers/handlers.js]: Throws have been changed to the same format as the rest (yes, it was an oversight). 

RE: Solutions branches, if you as maintainers want a different format for solutions I'm happy to acquiesce to that, I was just under the impression that solutions/mentor notes branches were the "done" thing. Hadn't seen foundersandcoders/master-reference#456.

Lock files are part of a node repo. npm explicitly tells you that they are supposed to be committed, my opinion would be that unless we as an organisation explicitly decide they are _not_ to be committed then we should conform to what npm suggest. As for "other stuff" idk what you mean, sorry.

File: step-1.5/public/index.js
Comment (marisid) [inline step-1.5/public/index.js]: Revert to ES5 `function () {}` instead of ES6 arrow function in front-end JS to avoid issues with browser compatibility. Also, plz make changes to the file in all the steps.

File: src/handlers/handlers.js
Comment (marisid) [reply src/handlers/handlers.js]: Inconsistent use of ES5 features (`var`, `function(request, response) {})` which you don't use later - is this on purpose so that you convert to ES6 together in the code-along? I would prefer to stick to ES6 throughout  (which we didn't do for FAC10 as I retrospectively believe we should). The same applies to the router file below.

**Review by marisid — CHANGES_REQUESTED** [review]
Hey @ZooeyMiller great job to add notes - I'm sure this will help you be on track on the day and remind yourself of the next step. Very detailed and descriptive. 

I've now added some suggestions in some parts of the code - plz have a look. 

Also, in general, I feel it's important to keep up with the trends and changes in JS. Great that you've converted the back-end JS in ES6 (with a few exceptions - see my comments) as a lot of great ES6 features are supported in Node. Could I also suggest to introduce `import` and `export default` to the students for importing and exporting modules using ES6 syntax? 

Regarding lock-files, as @eliascodes has mentioned, do we really need them in this workshop?

File: src/handlers/handlers.js
Comment (eliasmalik) [reply src/handlers/handlers.js]: Was just including the ref to foundersandcoders/master-reference#456 as reference, not expecting that you'd read it. Anyway, fair to say there's not yet consensus on the branch thing. I'd be happy to merge this into `mentor-walkthrough` for now, I just wanted to raise it with @jsms90 @des-des and @bradreeder.

I agree that lock files are part of a node project. I was asking whether we really needed to make each step-X directory a fully fledged node project; my understanding was that they were just reference examples. If that's true, we'd just need the HTML/CSS/JS. "other stuff" just means the `package.json`, lol.

**Comment by eliasmalik** [comment]
@marisid ES6modules will require transpilation. Do you think it's worth mentioning it here if they wont use it until (at the very earliest) their week 11 projects (and only if they choose a stack that includes babel)?

**Comment by ZooeyMiller** [comment]
@marisid and @eliascodes I really _like_ the idea of using `import` and `export default` and have used it a bit, but I don't feel comfortable with it to the point where I feel I could put it into this workshop and not do it a disservice.

File: src/server.js
Comment (ZooeyMiller) [reply src/server.js]: Ah, these are actually my original notes from the code-along! :sweat_smile: I should remove them completely, I believe this info is in the guide for mentors .md file

**Comment by ZooeyMiller** [comment]
All requested changes have been made.
Lock files are still in, but if it's blocking the merge then I can remove them. 
@eliascodes @marisid

**Comment by marisid** [comment]
@ZooeyMiller I wouldn't like you to talk about something you don't feel comfortable about. It was just a suggestion, as I think it is good-to-have knowledge when learning Node today. 

@eliascodes even if the students don't use a stack with a transpiler at some point (which I find unlikely in the longer term), I think it would be worth mentioning at the end of the workshop something along the lines: 

*On some occasions you may come across `import`/`export` instead of `require`/`module.exports` in some Node projects. This is basically the equivalent ES6 syntax that can be used in projects where your stack includes a transpiler. A transpiler (such as babel) is a tool to produce equivalent code in another syntax or language.* Which is basically what you've said :smile: 

Again, this is just a suggestion if you feel like the cohort won't be confused and the mentor feels comfortable to deliver.

**Review by marisid — APPROVED** [review]

**Review by eliasmalik — APPROVED** [review]

**Review by yvonne-liu — CHANGES_REQUESTED** [review]
I would like to fix some markdown issues before this is merged.

**Review by yvonne-liu — APPROVED** [review]

**Pull request merged**
